### ISSUE TITLE
BeforeAll and AfterAll

### ISSUE THREAD

**Issue by matkoch** [nuke-build/nuke#303]
https://github.com/nuke-build/nuke/issues/93#issuecomment-508933294

**Comment by matkoch** [comment]
@dennisdoomen I think I will refrain from implementing this. Is there any reason not to put the _Clean_  code into the constructor and control it via `--clean` parameter?

**Comment by dennisdoomen** [comment]
> The soft dependencies do help in this example, but just consider a large project with a lot of targets. In most of my projects, I have a single Default target that wants to run Clean first and then runs 10 other targets of which most could run in parallel if this feature ever landed. However, they all must run after Clean. Now, each target must know about the existence of the Clean target and have a After(Clean) clause added everywhere. I love Nuke, especially because the way dependencies work, but sometimes I just do want to specify some ordered dependencies in such a way I don't have to couple my targets together too much.

It's not just about the `Clean` target. I can't show you the example here because of confidential reasons, but I have several occasions where the order _is_ important. I _can_ add the targets that another target depends on to that target. I just want to be able to specify that _if_ a dependent target runs, it needs to run _before_ a specific target.

**Comment by dennisdoomen** [comment]
And one more argument for this is that the lack of this possibility is the biggest hurdle for people to move from PSake/Cake to Nuke.

**Comment by matkoch** [comment]
@dennisdoomen I’m not sure I understand. Putting the code of a `Clean` target into the constructor would have the exact same effect as defining it in a target and putting `BeforeAll` on it.

**Comment by dennisdoomen** [comment]
Let's say A does not directly depend on B, so B can run without A. But if A also runs (because of other dependencies, it must be run before B).

**Comment by matkoch** [comment]
That would be possible too. There’s a property `ExecutingTargets` and `InvokedTargets`.

**Comment by matkoch** [comment]
The problem that I have with this, as stated in the other comment, is that both implementations are not optimal. Plus, this feature request really only makes sense in the context of cleanup code, which is why I’m trying to find alternative ways.

**Comment by dennisdoomen** [comment]
> That would be possible too. There’s a property `ExecutingTargets` and `InvokedTargets`.

How? Can't find any docs on this.

**Comment by dennisdoomen** [comment]
> Putting the code of a `Clean` target into the constructor would have the exact same effect as defining it in a target and putting `BeforeAll` on it.

But that would mean it'll always run. If the build file is big enough, you don't want this to happen.

**Comment by dennisdoomen** [comment]
Anyway, I respect your work and opinions. I think adopting Nuke for open-source projects is going to be a nobrainer and a great step forwards, but IMO, moving an enterprise-class project from Cake to Nuke probably requires redesigning the entire build pipeline.

**Comment by matkoch** [comment]
> How? Can't find any docs on this.

http://www.nuke.build/docs/authoring-builds/predefined-parameters.html

> But that would mean it'll always run. If the build file is big enough, you don't want this to happen.

In my initial comment I said that this could be controlled with a simple parameter. Roughly like this:

```
[Parameter] bool Clean;

public Build()
{
    if (Clean)
    {
        // Clean code
    }
}
```

Note that we would have to reconsider how the build instance is initialized. So currently, this scenario is not _exactly_ supported.

**Comment by matkoch** [comment]
@dennisdoomen 

> moving an enterprise-class project from Cake to Nuke probably requires redesigning the entire build pipeline

Am I right you're talking about the `Setup` and `TearDown` part? According to https://cakebuild.net/docs/fundamentals/setup-and-teardown

**Comment by wgebczyk** [comment]
Hi,

I have issue to properly setup build script and it seems related to this issue.

When having such script:`

    Target Clean => _ => _
        .DependsOn(CleanIt)
        .Executes(() => { });

    Target Build => _ => _
        .After(Clean)
        .DependsOn(BuildIt)
        .Executes(() => { });

    Target CleanIt => _ => _
        .Executes(() => { Info("Clean It"); });

    Target BuildIt => _ => _
        .Executes(() => { Info("Build It"); });

    Target Rebuild => _ => _
        .DependsOn(Clean, Build)
        .Executes(() => { });
`
Will give the following output:
* Build It
* Clean It

It seems that After does not preserve order for child DependsOn targets.

Is this behavior intentional? Is this possible to define classic Clean, Build + Rebuild (Clean, Build) in nuke?

**Comment by matkoch** [comment]
Calling `nuke --plan` reveals the resulting dependency graph:

<img width="237" alt="image" src="https://user-images.githubusercontent.com/5005566/63353100-8493c980-c362-11e9-9ef4-80e74195c50e.png">

This behavior is intentional. If two targets have their order not specified by some kind of dependency, they can run in parallel. And that is exactly the behavior you get.

**Comment by matkoch** [comment]
Btw, I'm not exactly sure why you have multiple build and clean targets.

**Comment by wgebczyk** [comment]
Hey!

"why multiple build" -- this was just simplified script that I tried to create. Simply I do have Build target where it would depend on some smaller Build* targets: Build -> BuildProject1, BuildProject2, BuildPrepare, etc. Similar things with Cleanup. The smaller build targets would have inter dependencies as well.

It seems that it is not so obvious that task "after" dependency relates only to those tasks "Execute" and nothing else.

OK, might be the tool is not suit for my purposes. Is this still possible to define Clean+Build+Rebuild without defining "after" dependencies for each smaller build tasks?

**Comment by matkoch** [comment]
I think the documentation is specific on this. For instance:

<img width="867" alt="image" src="https://user-images.githubusercontent.com/5005566/63358054-bb6ddd80-c36a-11e9-8880-5201c0643fd0.png">

or

<img width="647" alt="image" src="https://user-images.githubusercontent.com/5005566/63358093-c9236300-c36a-11e9-9045-e1016de95b63.png">

And besides, it would be very confusing if there is _any_ chance for targets to run in parallel. If dependencies were to be inherited, then certainly not. Maybe there will be a time when it can be expressed in another way, but currently putting `After(Clean)` on all `BuildX` targets is required.

**Comment by wgebczyk** [comment]
Thanks for this explicit quote. Still term "target" might be understood as "target's Execute" or "whole target's definition".

That's good you have clear vision for your project. I hope you would stay with it and not give up to external influences :) Even though, this project will not fit my needs.

**Comment by matkoch** [comment]
I'm a bit surprised that this minor thing is the reason 🤔 Aren't you interested in running targets in parallel, when possible? If not okay, but otherwise your request is contradictory.

In any case, you can achieve what you want by applying an attribute similar as described here: https://github.com/nuke-build/nuke/issues/224#issuecomment-460463464

**Comment by wgebczyk** [comment]
Hey!

- I would avoid describing things as minor, etc. For some might be minor, for others way more significant.
- For my pet project I would not get too much from parallelization - I have 3 big TypeScript projects and there are some intermediate step to prepare those TS ones. Paralellization would help for my daily work where are about main 400 projects + a lot of smaller tasks for full build -- but that's another story.

I've gave this issue bit more time and played a bit with it. I've realized that I'm trying to mimic some other concept with nuke's Targets and that makes really no sense. For case of "Clean+Build+Rebuild" (CBR), we have some set of project that can be compiled and/or cleaned. We are executing all of them as Build or as Clean for some consistent period of time. The set of those tasks is same or almost the same, so makes sense to re-use them for Build and Clean.

I've created this Task method that works for me, but... looks and is used rather ugly.
I would prefer to use more fluent approach, but I do not know how to extend nuke.
Would be nice to have some extension method to ITargetDefinition, but it must be static and Clean is instance field.

`

    [Parameter] bool    Clean              = false;

    Target CommonTask => _ => Task(_,
        whenBuild: () => {
            Info("[CommonTask] whenBuild");
        },
        whenClean: () => {
            Info("[CommonTask] whenClean");
        }
    );

    Target Project1PrepareTask => _ => Task(_
        .DependsOn(CommonTask),
        whenBuild: () => {
            Info("[Project1PrepareTask] whenBuild -- use things from [CommonTask]");
        }
    );

    Target Project1Task => _ => Task(_
        .DependsOn(Project1PrepareTask),
        whenBuild: () => {
            Info("[Project1Task] whenBuild - use things from [Project1PrepareTask]");
        },
        whenClean: () => {
            Info("[Project1Task] whenClean");
        }
    );

    Target Project2PrepareTask => _ => Task(_
        .DependsOn(CommonTask),
        whenBuild: () => {
            Info("[Project2PrepareTask] whenBuild -- use things from [CommonTask]");
        }
    );

    Target Project2Task => _ => Task(_
        .DependsOn(Project1Task, Project2PrepareTask),
        whenBuild: () => {
            Info("[Project2Task] whenBuild - use things from [Project2PrepareTask] and reference [Project1Task]");
        },
        whenClean: () => {
            Info("[Project2Task] whenClean");
        }
    );

    Target Project3Task => _ => Task(_
        .DependsOn(Project1Task, Project2Task),
        whenBuild: () => {
            Info("[Project3Task] whenBuild - reference [Project1Task] and [Project2Task]");
        }
    );

    Target Build => _ => Task(_
        .DependsOn(Project3Task)
    );

    private ITargetDefinition Task(ITargetDefinition target, Action whenBuild = null, Action whenClean = null)
    {
        return target
            .Executes(() => {
                if (!Clean) whenBuild?.Invoke();
            })
            .Executes(() => {
                if (Clean) whenClean?.Invoke();
            });
    }

**Comment by RLittlesII** [comment]
@wgebczyk https://github.com/nuke-build/nuke/issues/303#issuecomment-523165607 seems like a long way around just using [Conditions](https://www.nuke.build/docs/authoring-builds/fundamentals.html#static-conditions--dependency-behavior) to dictate how your build will execute.

I have gone as far in Cake and in Nuke to create higher level Task/Targets that just surface sub task/targets with criteria.  The main down side to this approach is you end up with a lot of skipped task/targets in your completed output.

I am late to this party, but does this sound like what your attempting to accomplish?

**Comment by dennisdoomen** [comment]
I've been using build scripts for more than two decades (when I was still writing C++), and have since worked with simple cmd-line scripts, Powershell scripts, PSake, Cake and now Nuke. I love Nuke for many reasons, including its dependency model that will eventually support parallel execution of targets. But I do feel that moving a complicated Cake script to Nuke is not going to be worth it. The dependency model is so different from Cake's, and having to jump through all kinds of hoops without having to completely redesign your target graph is not going to cut it to convince large teams like mine. If you start from scratch, Nuke is my absolute first choice though.

**Comment by wgebczyk** [comment]
@RLittlesII I've considered conditions, but haven't found better way to use that, that by duplicating  whole Target with the only difference on condition and execute. That looks ugly for me and more hassle that what I have now. Have you found better way of using conditions? Could you put here some snippet?

**Comment by matkoch** [comment]
@wgebczyk 

> I would avoid describing things as minor, etc. For some might be minor, for others way more significant.

Can't see how it has a significant impact, sorry. The workaround is to add `Before`/`After` more often. Looking at other reports and planned features it _is_ - with all respect - a minor issue. Also whether or not your pet project benefits from parallelization is not really of matter for this discussion. We aim to solve _common/abstract_ issues in build automation, but we won't go a route that only serves individual needs or blocks future ideas. I hope for your understanding.

> Would be nice to have some extension method to ITargetDefinition, but it must be static and Clean is instance field.

Then declare it `static` ?

**Comment by wgebczyk** [comment]
@matkoch I appologize for bringing irrelevant matter to this discussion. I was mislead by your question why I'm not interested in running in parallel.

This means only that your tool targets big, trendy cases, have strong opinion about throwing more after/before to solve issues. That's ok! However, I think that this tool is not timed for small project outside of your envisioned usage.

I believe that high code quality should be expected from sources, tests and all other scripts including build ones. And throwing more this or that to solve issues is not good long term solution, even for small projects.

Wish you all the best with big fishes! :)

**Comment by matkoch** [comment]
Thanks, and thanks for your time, although I honestly still don't understand the issue.

**Comment by matkoch** [comment]
Closed since not much happened in the last 2 years. Meanwhile I'm much against the idea, but i will probably introduce another solution eventually. Nothing planned though.
